src/server/worker/datum/iterator.go

```diff
@@ -453,7 +455,12 @@ func (ds *datumStream) Priority() int {
 }
 
 // NewIterator creates a new datum iterator.
-func NewIterator(pachClient *client.APIClient, input *pps.Input) (Iterator, error) {
+func NewIterator(pachClient *client.APIClient, input *pps.Input) (iterator Iterator, retErr error) {
+	defer func() {
+		if retErr == nil {
+			iterator = newIndexIterator(iterator)
+		}
+	}()
```

Comment:
Using a deferred function here seems like it could potentially confuse someone and might make debugging hard. I think it'd make more sense to define iterator as a variable in the function. Or just add `newIndexIterator` to each of the return sites.

Reply:
I disagree with this. We use this pattern in quite a few places in the codebase and I think it improves the readability and maintainability. On the readability front, it separates a common transform to be applied to the results of a function from the core functionality of the function. On the maintainability front, changes only need to take into account the core functionality or the common transform, not both. If these two ideas were combined, it becomes much easier to make an incomplete change (i.e. missing wrapping a new iterator with `newIndexIterator`).

Reply:
I like the pattern of declaring an interface variable and repeatedly writing to it.  I think this is what JD meant by "variable in the function"

```go
var iterator Iterator
// setup the base iterator with a constructor, erroring out if needed.
iterator = newIndexIterator(iterator)
iterator = newOtherKindOfIteratorLayer(iterator)
iterator = newLastHigherOrderIterator(iterator)
return iterator
```

Reply:
IMO defer is sort of an anti-pattern outside RAII.

Reply:
Okay, that makes sense. I guess it is unnecessary to handle this in a defer since it doesn't apply in the error case and if we add more wrappers it would end up looking like that anyways or be more defers which seems bad. We probably want to make a similar change in the `obj` package for the checked / uniform client.

Reply:
Sounds like we have a good solution here. I think I generally agree with Brendon that defers should basically be reserved for code that needs to run after you return an error.